### PR TITLE
Changelog v1.4

### PR DESCRIPTION
**Pull request by curquiza**
Fixes https://github.com/meilisearch/meilisearch/issues/3880

### CODE REVIEW

File: meilisearch-release-changelog/v1.4.0.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
+# v1.4.0 release changelogs
+
+<!-- The following line should NOT be put in the official release changelogs -->
+⚠️ Since this is a release candidate (RC), we do NOT recommend using it in a production environment. Is something not working as expected? We welcome [bug reports](https://github.com/meilisearch/meilisearch/issues/new/choose) and [feedback about new features](https://github.com/meilisearch/product/discussions).
+
+Meilisearch v1.4 introduces features to improve the relevancy of the Meilisearch results by adding three news settings: `nonSeparatorTokens`, `separatorTokens` and `dictionary`.
```
Comment (guimachiavelli) [inline meilisearch-release-changelog/v1.4.0.md:6]:
```suggestion
Meilisearch v1.4 improves the search result relevancy by introducing three new settings: `nonSeparatorTokens`, `separatorTokens` and `dictionary`.
```

File: meilisearch-release-changelog/v1.4.0.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+<!-- The following lines should ONLY be put in the official release changelogs -->
+🧰 All official Meilisearch integrations (including SDKs, clients, and other tools) are compatible with this Meilisearch release. Integration deployment happens between 4 to 48 hours after a new version becomes available.
+
+Some SDKs might not include all new features—consult the project repository for detailed information. Is a feature you need missing from your chosen SDK? Create an issue letting us know you need it, or, for open-source karma points, open a PR implementing it (we'll love you for that ❤️).
+
+# New features and improvements 🔥
+
+## Customize text separators
+
+The Meilisearch word segmentation uses a default list of separators to segment words. This default list might not sometimes fit your use case: the `nonSeparatorTokens`/`separatorTokens`` setting allows you to remove/add some tokens from/to the default list of separators.
```
Comment (guimachiavelli) [inline meilisearch-release-changelog/v1.4.0.md:17]:
```suggestion
Meilisearch word segmentation uses a list of characters to separate one word for another. The `nonSeparatorTokens` and `separatorTokens` index settings allow you to configure this to better fit your dataset.
```

File: meilisearch-release-changelog/v1.4.0.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  --data-binary '["@", "#", "&"]'
+```
+
+To add tokens:
+```bash
+curl \
+  -X PUT 'http://localhost:7700/indexes/articles/settings/separator-tokens' \
+  -H 'Content-Type: application/json'  \
+  --data-binary '["&sect;", "&sep"]'
+```
```
Comment (guimachiavelli) [inline meilisearch-release-changelog/v1.4.0.md:34]: I find it a bit easier to understand this if we begin with the setting with the simpler name (`separator-tokens`), then progress to the more complex name (`*non*-separator-tokens`)

````suggestion
Add a character to the `separator-tokens` list to use it as a word separator:

```bash
curl \
  -X PUT 'http://localhost:7700/indexes/articles/settings/separator-tokens' \
  -H 'Content-Type: application/json'  \
  --data-binary '["&sect;", "&sep"]'
```

Add a character to the `non-separator-tokens` list when you don't want Meilisearch to use it to separate words:

```bash
curl \
  -X PUT 'http://localhost:7700/indexes/articles/settings/non-separator-tokens' \
  -H 'Content-Type: application/json'  \
  --data-binary '["@", "#", "&"]'
```
````

File: meilisearch-release-changelog/v1.4.0.md [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+Done by @ManyTheFish in #3946
+
+## Load a user-defined dictionary
```
Comment (guimachiavelli) [inline meilisearch-release-changelog/v1.4.0.md:38]:
```suggestion
## Load user-defined dictionaries
```

File: meilisearch-release-changelog/v1.4.0.md [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+The Meilisearch word segmentation relies on separators and language-based dictionaries to segment words. This segmentation could be more accurate on technical or use-case-specific vocabulary (like "G/Box" to say "Gear Box") or proper nouns (like "J. R. R." when parsing "J. R. R. Tolkien"). The `dictionary` setting improves this accuracy by defining a list of words that would be segmented as described in the list.
```
Comment (guimachiavelli) [inline meilisearch-release-changelog/v1.4.0.md:40]:
```suggestion
Meilisearch word segmentation also relies on language-based dictionaries to segment words. Use the `dictionary` index setting to expand the default dictionaries and improve accuracy when working with datasets using domain-specific terms:
```

File: meilisearch-release-changelog/v1.4.0.md [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+```bash
+curl \
+  -X PUT 'http://localhost:7700/indexes/articles/settings/dictionary' \
+  -H 'Content-Type: application/json'  \
+  --data-binary '["J. R. R.", "J.R.R."]'
+```
+
+These feature synergies well with the `stopWords` or the `synonyms` settings allowing to segment words and correctly retrieve what you expect:
```
Comment (guimachiavelli) [inline meilisearch-release-changelog/v1.4.0.md:49]:
```suggestion
This feature can be used together with the `stopWords` and `synonyms` index settings:
```

File: meilisearch-release-changelog/v1.4.0.md [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+```
+
+Done by @ManyTheFish in #3946
+
+## Other improvements
+TBD
+
+# Fixes 🐞
+
+* Fix display of synonyms (#3946) @ManyTheFish
```
Comment (guimachiavelli) [inline meilisearch-release-changelog/v1.4.0.md:69]:
```suggestion
* Fix synonyms display (#3946) @ManyTheFish
```

File: meilisearch-release-changelog/v1.4.0.md [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+  * Upgrade Meilisearch dependencies (#3987) @ManyTheFish
+* CIs and tests
+  * Fix benchmarks CI (#3963) @irevoire
+  * Removed unnecessary borrow call that failed nightly tests (#3990) @JannisK89
+  * Expose a Puffin server to profile the indexing process (#3913) @Kerollmops
+  * Update and fix the Test Suite CI (#3918) @Kerollmops
+  * Improve test suite CI for manual trigger events (#3989) @curquiza
+* Documentation
+  * Update README.md (#3895 and #3932) @ferdi05 & @Strift
+* Misc
```
Comment (guimachiavelli) [inline meilisearch-release-changelog/v1.4.0.md:84]: Seems unnecessary to have sublists here, especially as the last one is a repetition of the section title (that is, `# Misc`).

```suggestion
  * Upgrade Meilisearch dependencies (#3987) @ManyTheFish
  * Upgrade CI dependencies (#3871, #3969 and #3968)
  * Fix CI benchmarks (#3963) @irevoire
  * Removed `borrow` call causing failed nightly tests (#3990) @JannisK89
  * Expose Puffin server to profile the indexing process (#3913) @Kerollmops
  * Update and fix the CI Test Suite (#3918) @Kerollmops
  * Improve CI test suite for manual trigger events (#3989) @curquiza
  * Update README.md (#3895 and #3932) @ferdi05 & @Strift
```

**Review by guimachiavelli — CHANGES_REQUESTED** [review]

File: meilisearch-release-changelog/v1.4.0.md (hunk 2 above)
Comment (Kerollmops) [reply meilisearch-release-changelog/v1.4.0.md:17]:
```suggestion
Meilisearch word segmentation uses a list of characters to separate one word from another. The `nonSeparatorTokens` and `separatorTokens` index settings allow you to configure this to better fit your dataset.
```

**Comment by Kerollmops** [comment]
Thank you @guimachiavelli for your review I accepted most of it 😄
If you got time at some point, I added things in my last commit 🙏
Have a nice day 🏝️

File: meilisearch-release-changelog/v1.4.0.md [hunk 9]
```diff
@@ -63,14 +63,24 @@ This feature can be used together with the `stopWords` and `synonyms` index sett
...
+```
+
+Done by @vivek-26 in #3945
 
 # Fixes 🐞
 
-* Fix synonyms display (#3946) @ManyTheFish
+  * Fix synonyms display (#3946) @ManyTheFish
+  * Accept the `null` JSON value as a value of the `_vectors` field (#3998) @Kerollmops
+  * Fix the ranking rule by temporarily disabling an assert in the bucket sort algorithm (#4013) @Kerollmops
```
Comment (curquiza) [inline meilisearch-release-changelog/v1.4.0.md:79]: @Kerollmops looks like it's v1.3.2 PRs, I mean, they are in the v1.3.2 Milestones. Where is the error, in this changelogs or the milestones linked to the PRs?

File: meilisearch-release-changelog/v1.4.0.md [hunk 10]
```diff
@@ -63,14 +63,24 @@ This feature can be used together with the `stopWords` and `synonyms` index sett
...
 
 # Misc
 
+  * Bump charabia to 0.8.3 (#4008) @Kerollmops
```
Comment (curquiza) [inline meilisearch-release-changelog/v1.4.0.md:83]: @Kerollmops same looks like it's not a PR for v1.4.0 can you confirm

File: meilisearch-release-changelog/v1.4.0.md [hunk 11]
```diff
@@ -79,8 +89,7 @@ TBD
   * Update and fix the CI Test Suite (#3918) @Kerollmops
   * Improve CI test suite for manual trigger events (#3989) @curquiza
   * Update README.md (#3895 and #3932) @ferdi05 & @Strift
-  * TBD
 
 ❤️ Thanks again to our external contributors:
-- [Meilisearch](https://github.com/meilisearch/meilisearch): @JannisK89
-- [Charabia](https://github.com/meilisearch/charabia): TBD
+- [Meilisearch](https://github.com/meilisearch/meilisearch): @JannisK89, @vivek-26
+- [Charabia](https://github.com/meilisearch/charabia): @vvv
```
[the remark this answers is not in the document]
Comment (Kerollmops) [reply meilisearch-release-changelog/v1.4.0.md:96]: I'm really confused here. I don't know where I found a contribution from him 😔 Sorry!

File: meilisearch-release-changelog/v1.4.0.md (hunk 9 above)
Comment (Kerollmops) [reply meilisearch-release-changelog/v1.4.0.md:79]: Indeed, it's a mistake. Those changes are in v1.3.2. Sorry for that!

File: meilisearch-release-changelog/v1.4.0.md (hunk 10 above)
Comment (Kerollmops) [reply meilisearch-release-changelog/v1.4.0.md:83]: Nope, it has been done in v1.3.2. I don't know what I did there.

File: meilisearch-release-changelog/v1.4.0.md [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+* Fix synonyms display (#3946) @ManyTheFish
+* Fix thai synonyms (#4033) @Kerollmops
+* Fix synonyms with separators (#3994) @ManyTheFish
+* Fix the swap index tasks (#4041) @irevoire
+* Fix highlighting bug when searching for a phrase with cropping (#4028) @vivek-26
+* Empty arrays/objects now return empty instead of null (#3997) @dogukanakkaya
+* Fix the stats of the documents deletion by filter (#4053) @irevoire
+* `\` can now be used at the end of the `filter` search parameter values (#4038 and #4043) @Kerollmops
+
+## 🔴 Usage breaking change following `filter` bug fix
```
Comment (curquiza) [inline meilisearch-release-changelog/v1.4.0.md:83]: @Kerollmops @irevoire 😇

File: meilisearch-release-changelog/v1.4.0.md [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+    "path": "my\\test\\path\\"
+  }
+]
+```
+
+(`\\` escaping here is about JSON escaping)
+
+Before v1.4.0, if you wanted to filter on document 2, the filters `path = "my\\test\\path\\"` and `path = "my\\test\\path\\\"` led to an error.
+
+Now, in v1.4.0, you can express all the possible filter expressions, no matter if the filter contains `\` or not. But to reach this, **you have escape all the `\` characters in your filter**.
```
Comment (curquiza) [inline meilisearch-release-changelog/v1.4.0.md:108]:
```suggestion
Now, in v1.4.0, you can express all the possible filter expressions, no matter if the filter contains `\` or not. But to reach this, **you have to escape all the `\` characters in your filter**.
```

File: meilisearch-release-changelog/v1.4.0.md [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+]
+```
+
+(`\\` escaping here is about JSON escaping)
+
+Before v1.4.0, if you wanted to filter on document 2, the filters `path = "my\\test\\path\\"` and `path = "my\\test\\path\\\"` led to an error.
+
+Now, in v1.4.0, you can express all the possible filter expressions, no matter if the filter contains `\` or not. But to reach this, **you have to escape all the `\` characters in your filter**.
+
+Following our example, to finally succeed in filtering on document 2, the filter should be: `path = "my\\\\test\\\\path\\\\"`.
```
Comment (Kerollmops) [inline meilisearch-release-changelog/v1.4.0.md:110]: On this line, the JSON and filtering escaping are visible:
`\\\\` is giving the filter a `\\` and the filter gives us a single `\`.

File: meilisearch-release-changelog/v1.4.0.md [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+
+👉 If you come from v1.3.X or before, and you used `\` in `filter`, here is the change to apply: following our example, to filter on document 1, before v1.4.0, you could apply the filter `path = "my\\test\\path"`. Now, from v1.4.0, you have to escape the `\` and apply the following filter: `path = "my\\\\test\\\\path"`.
+Two escapings are applied: the JSON escaping and the Meilisearch filter escaping. The Meilisearch filter transforms the `\\\\` into `\\` and the JSON parsing ends up with a single `\`.
```
Comment (Kerollmops) [inline meilisearch-release-changelog/v1.4.0.md:113]: Nope that the other way around:
```suggestion
Two escapings are applied: the JSON escaping and the Meilisearch filter escaping. The JSON unescaping transforms the `\\\\` into `\\` and the Meilisearch filter unescaping ends up with a single `\`.
```

**Review by Kerollmops — CHANGES_REQUESTED** [review]

**Comment by curquiza** [comment]
Release done!

**Pull request merged**